README.md
# project-pan

[![CI](https://github.com/yetanalytics/project-pan/actions/workflows/main.yml/badge.svg)](https://github.com/yetanalytics/project-pan/actions/workflows/main.yml)

A Clojure library for validating xAPI Profiles, according to the [xAPI Profile specification](https://github.com/adlnet/xapi-profiles).

## Installation

Add the following to the `:deps` map in your `deps.edn` file:

```clojure
com.yetanalytics/project-pan {:mvn/version "0.5.4"
                              :exclusions [org.clojure/clojure
                                           org.clojure/clojurescript]}
```

## Usage

To use the library to validate a whole Profile, call `validate-profile` in the `pan` namespace. This method takes in an entire Profile as an EDN data structure and prints either a success message on success (obviously) or an error message on failure.

Similarly, to validate a collection of Profiles, call `validate-profile-coll` on them. Each Profile in the collection will be able to reference Concepts, Templates, and Patterns from the other Profiles.

Arguments may be supplied for different levels of validation strictness, which are listed as follows:
- `:syntax?` - Basic validation; check only the types of properties and simple syntax of all Profile objects. Default `true`.
- `:ids?` - Validate the correctness of all object and versioning IDs (the `id` and `inScheme` properties). Validate that all IDs are distinct and that all `inScheme` values correspond to valid Profile version IDs. If multiple Profiles are involved, all IDs are checked that they are distinct among _all_ Profiles (not just the Profile it is a part of).\*
- `:multi-version?` - A flag for the `:ids?` keyword arg; if `false` then ID validation will check that all objects have the same `inScheme` value. If `true`, then in addition to the usual ID checking, validation that Template and Pattern IDs change when their properties (other than `inScheme`, `prefLabel`, and `scopeNote`) change is active. Default `false`.
- `:relations?` - Validate that all relations between Profile objects are valid. These relations are given by IRIs and include the following:
    - the broader, narrower, related, broadMatch, narrowMatch, relatedMatch and exactMatch properties for Verbs, Activity Types and Attachment Usage Types.
    - the `recommendedActivityTypes` property for Activity Extensions
    - the `recommendedVerbs` property for Context and Result Extensions
    - Determining Properties, `objectStatementRefTemplate` property and the `contextStatementRefTemplate` properties for Statement Templates.
    - `sequence`, `alternates`, `optional`, `oneOrMore` and `zeroOrMore` properties for Patterns.
- `:concept-rels?`, `:template-rels?`, `:pattern-rels?` - Similar to `:relations?`, except that each only validates relations specific to Concept, Statement Template, and Pattern properties, respectively. Each are default `false`, and are overridden when `:relations?` are `true`.
- `:contexts?` - Validate that all instances of `@context` resolve to valid JSON-LD contexts and that they allow all properties to expand out to absolute IRIs during JSON-LD processing. The `@context` property is always found in the Profile metadata and in Activity Definitions, though they can also be found in Extensions for said Activity Definitions.

Other keyword arguments include:
- `:external-profiles` - Extra Profiles from which to reference Concepts, Statement Templates, and Patterns from. Unlike the Profiles passed into `validate-profile-coll`, these Profiles are not validated (though will be treated as part of the same scope in terms of ID distinctiveness validation).
- `:external-contexts` - Extra `@context` values (other than the xAPI Profile and Activity contexts) that `@context` IRIs in `profile` can reference during context validation. During multi-profile validation, all Profiles refer to the same global `:external-contexts` map. Default `{}`.
- `:result` - This value affects the format of the result type.
    - `:print` prints an [Expound](https://github.com/bhb/expound)-generated error string to standard output.
    - `:string` returns the aforementioned error string (or a vector of them, in the case of `validate-profile-coll`).
    - `:type-string` returns a `{:type string}` map, where `:type` is a keyword and `string` is an error message specific to that type of error. `:type` can be one of the following:
      - `:syntax-errors` for basic validation errors
      - `:id-errors` for ID errors
      - `:in-scheme-errors` for inScheme errors
      - `:concept-edge-errors` for errors between Concept relations
      - `:template-edge-errors` for errors between Template relations
      - `:pattern-edge-errors` for errors between Pattern relations
      - `:pattern-cycle-errors` for cycles in the Pattern relation graph
      - `:context-errors` for errors relating to JSON-LD context
    - `:type-path-string` returns a `{:type {path string}}` map, where `path` is the `assoc-in` vector needed to retrieve the erroneous value, and `string` is an error message specific to that error location.
   - `:spec-error-data` - Return a `{:type spec-error-data}` map, with each value being the map generated by Clojure spec. Default.
- `:error-msg-opts` - Takes an opt map that affects how error messages are formed and printed. Currently only takes `:print-objects?`, which when `true` (default) displays the entire object in error messages, and if `false` only displays IDs.

### Additional API Functions

The `validate-object` function is provided in order to provide a way to validate individual Concepts, Templates, and Patterns without having to go deeper than the top level API. The `:type` and `:result` keyword args are used to fix the expected spec and to affect the function result, respectively; more details can be found in the docstring.

The `get-external-iris` function is provided in the API in order to retrieve IRI values that refer to external objects, JSON-LD contexts, etc. (i.e. objects that do _not_ exist in the Profile). This allows the user to more easily retrieve external Profiles and contexts in their application from the Internet or their data store.

The `json-profile->edn` function is provided in the API to provide for convenient coercion of JSON profile strings into an EDN format that Pan recognizes.

Besides validating whole Profiles, you can also use library methods and specs to validate parts of Profiles, such as individual  Concepts, Templates and Patterns).

## TODO

- Be able to access external links:
    - Non-inline `schema` values are given by IRIs that link out to external JSON Schemas.
    - Except for the two w3id contexts given by the Profile spec, any and all `@context` values need to be given by external links.
- Log errors to somewhere instead of printing them out.
- More graceful exception handling.
- Any bugs that need to be fixed (natch).

## License

Copyright © 2019-2023 Yet Analytics, Inc.

Project Pan is licensed under the Apache License, Version 2.0. See [LICENSE](LICENSE) for the full license text


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 0.5.4 - 2023-05-02
- Fix bug in Pattern cycle detection where non-cyclic paths were incorrectly flagged as cycles.

## 0.5.3 - 2022-10-24
- Update GitHub CD and CI to remove deprecation warnings.

## 0.5.2 - 2022-09-28
- Decouple Pan's `::activityDefinition` spec from that of the xapi-schema lib.
- Change the JSON parse function to not keywordize keys containing `:`.
  - This change mainly affects IRIs, such as Activity extension keys.
- Move all `clojure.test.check.generators` requires to test namespaces.
  - Downstream cljs applications no longer have to include the `test.check` lib.

## 0.5.1 - 2022-07-21
- Remove Loom dependency.
- Add instrumentation to graph util functions.

## 0.5.0 - 2022-06-14
- Relax ID validation constraints to allow different versions of the same Profile to reuse IDs in certain cases.
- Add a `:multi-version?` flag to validation functions to toggle whether multiple versions can exist in a single Profile object or not.
- Add the `validate-object` function to the top-level API to validate individual Concepts, Templates, and Patterns.
- Add value generation capabilities for Profile syntax specs, including `::concept`, `::template`, `::pattern`, and `::profile`.
- Fix bugs when validating Pattern relations across multiple Profiles. (See [#39](https://github.com/yetanalytics/project-pan/pull/39) for details.)
- Fix bugs where external `contextCategoryActivityType` IRIs were not properly validated. (See [#40](https://github.com/yetanalytics/project-pan/pull/40) for details.)

## 0.4.3 - 2022-03-04
- Fix bug where outgoing edges of profile-external StatementTemplates were incorrectly included in the Pattern graph.

## 0.4.2 - 2022-03-03
- Add `:concept-rels?`, `:template-rels?`, and `:pattern-rels?`, keyword arguments.

## 0.4.1 - 2022-02-25
- Add `json-profile->edn` API function.

## 0.4.0 - 2022-02-25
- Add multi-profile support
  - Extra API function `validate-profile-coll` that accepts a collection of Profiles to validate.
  - `:extra-profiles` keyword arg to reference Profiles with external Concepts, Templates, and Patterns.
  - `:extra-contexts` keyword arg to reference external, non-default JSON-LD contexts.
- Replace `:print-err?` keyword arg to `:result`, with options `:spec-error-data`, `:type-path-string`, `:path-string`, `:string`, and `:print`.
- Overhaul internal logic for identifier, graph and context validation.
- Rename various internal namespaces.

See [Pull Requests #23 to #27](https://github.com/yetanalytics/project-pan/pulls?q=is%3Apr+is%3Aclosed) for more details.

## 0.3.0 - 2021-04-07
- Upgrade to latest Expound version and clean up internal error logic.
- Made error messages more consistent across different error types.
- Rename certain spec names and simplify graph validation functions.
- Remove JSON Schema-related deps - no more downstream downloading of jitpack!
- Updated xapi-schema dep to reflect latest fixes.

## 0.2.0 - 2021-02-01
- Add ClojureScript compatibility.
- Singleton sub-Pattern IDs now MUST be IRIs, not object maps.

## 0.1.0 - 2019-09-12
- Initial alpha release.


src/com/yetanalytics/errors.clj
(ns com.yetanalytics.errors
  (:require [clojure.spec.alpha :as s]
            [clojure.core.match :as m]
            [clojure.string :as string]
            [expound.alpha :as exp]
            [expound.problems :as exp-prob]
            [expound.printer :as exp-print]
            [com.yetanalytics.axioms :as ax]
            [com.yetanalytics.context :as ctx]
            [com.yetanalytics.identifiers :as id]
            [com.yetanalytics.graph :as g]
            [com.yetanalytics.objects.concept :as c]
            [com.yetanalytics.objects.template :as t]
            [com.yetanalytics.objects.pattern :as p]
            [com.yetanalytics.objects.concepts.extensions.activity :as ae]
            [com.yetanalytics.objects.concepts.extensions.context :as ce]
            [com.yetanalytics.objects.concepts.extensions.result :as re]
            [com.yetanalytics.objects.concepts.activities :as act]
            [com.yetanalytics.util :as u]
            [xapi-schema.spec :as xs]))

;; TODO: Expound can act pretty funny sometimes and we have to use some hacks
;; to avoid said funniness. Look into creating an in-house error message lib.

;; Error map reference:
;; :path = Path of map + spec keywords
;; :pred = Spec predicate
;; :val = Value being specced
;; :via = Path of spec keywords
;; :in = Path of map keys + array entry numbers

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Syntax spec messages
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(exp/defmsg ::ax/boolean "should be a boolean")
(exp/defmsg ::ax/string "should be a non-empty string")
(exp/defmsg ::ax/timestamp "should be a valid timestamp")
(exp/defmsg ::ax/language-map "should be a valid language map")
(exp/defmsg ::ax/language-tag "should be a valid language tag")
(exp/defmsg ::ax/lang-map-string "should be a valid string")
(exp/defmsg ::ax/media-type "should be a RFC 2046 media type")
(exp/defmsg ::ax/json-path "should be a valid JSONPath string")
(exp/defmsg ::ax/json-path "should be a valid JSON schema")
(exp/defmsg ::ax/iri "should be a valid IRI")
(exp/defmsg ::ax/irl "should be a valid IRL")
(exp/defmsg ::ax/uri "should be a valid URI")
(exp/defmsg ::ax/url "should be a valid URL")
(exp/defmsg ::ax/array-of-iri "should be an array of IRIs")
(exp/defmsg ::ax/array-of-uri "should be an array of URIs")

(exp/defmsg ::u/inline-or-iri "cannot contain both linked and inline JSON schema")
(exp/defmsg ::u/related-only-deprecated "related can only be used on deprecated concepts")
(exp/defmsg ::ae/no-recommended-verbs "only allowed on context or result extensions")
(exp/defmsg ::ce/no-recommended-activity-types "only allowed on activity extensions")
(exp/defmsg ::re/no-recommended-activity-types "only allowed on activity extensions")
(exp/defmsg ::act/activityDefinition "invalid activity definition")

(exp/defmsg ::t/type-or-reference "cannot contain both objectActivityType and objectStatementRefTemplate")

(exp/defmsg ::p/pattern-clause "pattern contains too many properties")
(exp/defmsg ::p/is-primary-true "primary is not true")
(exp/defmsg ::p/is-primary-false "primary is not false nor nil")

;; ID spec messages
(exp/defmsg ::id/one-count "the id value is not unique")
(exp/defmsg ::id/in-scheme "the inScheme value is not a valid version ID")

;; Graph spec messages
(exp/defmsg ::g/not-self-loop "object cannot refer to itself")

(exp/defmsg ::c/valid-dest "linked concept does not exist")
(exp/defmsg ::c/relatable-src "should be type: \"ActivityType\", \"AttachmentUsageType\" or \"Verb\"")
(exp/defmsg ::c/relatable-dest "should link to type: \"ActivityType\", \"AttachmentUsageType\" or \"Verb\"")
(exp/defmsg ::c/aext-src "should be type: \"ActivityExtension\"")
(exp/defmsg ::c/crext-src "should be type: \"ContextExtension\" or \"ResultExtension\"")
(exp/defmsg ::c/at-dest "should link to type: \"ActivityType\"")
(exp/defmsg ::c/verb-dest "should link to type: \"Verb\"")
(exp/defmsg ::c/same-concepts "the concepts are not the same type")
(exp/defmsg ::c/same-version "inScheme values do not match")

(exp/defmsg ::t/template-src "should be type: \"StatementTemplate\"")
(exp/defmsg ::t/valid-dest "linked concept or template does not exist")
(exp/defmsg ::t/verb-dest "should link to type: \"Verb\"")
(exp/defmsg ::t/at-dest "should link to type: \"ActivityType\"")
(exp/defmsg ::t/aut-dest "should link to type: \"AttachmentUsageType\"")
(exp/defmsg ::t/template-dest "should link to type: \"StatementTemplate\"")
(exp/defmsg ::t/same-version "inScheme values do not match")

(exp/defmsg ::p/valid-dest "linked template or pattern does not exist")
(exp/defmsg ::p/pattern-src "should be type: \"Pattern\"")
(exp/defmsg ::p/pattern-dest "should link to type: \"Pattern\"")
(exp/defmsg ::p/template-dest "should link to type: \"StatementTemplate\"")
(exp/defmsg ::p/non-opt-dest "alternate pattern cannot contain an optional or zeroOrMore pattern")
(exp/defmsg ::p/singleton-src "primary sequence pattern has multiple links")
(exp/defmsg ::p/not-singleton-src "sequence pattern must have at least two links")
(exp/defmsg ::p/primary-pattern "pattern is not primary")
(exp/defmsg ::p/zero-indegree-src "pattern must not be used elsewhere")

(exp/defmsg ::p/singleton-scc "cyclical reference detected")

;; Context errors
(exp/defmsg ::ctx/simple-term-def "simple term definition does not have valid prefix")
(exp/defmsg ::ctx/expanded-term-def "expanded term definition does not have valid prefix")

(exp/defmsg ::ctx/contexed-key "key cannot be expanded into absolute IRI")
(exp/defmsg ::ctx/is-at-context "key is not @context")

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Value display 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn strv
  "Short for 'stringify-value.' Like str or pr-str, but makes special cases
  for nils and keywords."
  [value]
  (cond
    (nil? value) "nil"
    (= value :_context) "@context"
    (keyword? value) (name value)
    :else (pr-str value)))

(defn pluralize
  "Make the word plural based off of a count if needed, by adding a plural s
  at the end."
  [word cnt]
  (if (< 1 cnt) (str word "s") word))

;; Needed to get around Issue #110 in Expound
(defn default-printer
  "Return a default printer with :show-valid-values? and :print-specs? both
  preset to false. Also truncate lines that only contain elipses."
  [explain-data]
  (let [unformed-str
        (with-out-str ;; Pretty hacky, but it works
          ((exp/custom-printer {:show-valid-values? false
                                :print-specs? false}) explain-data))
        formed-str
        (string/replace unformed-str #"(?<=\n)\s+\.\.\.\n" "")]
    (print formed-str)))

;; TODO Current default-printer function is quite unhelpful in locating errors.
;; May look into this again.
#_(defn value-str-def
    "Custom value string for syntax validation error messages. Takes the form:
  > Invalid value: <value>
  > At path: profile -> <key1> -> <key2>"
    [_ form path value]
    (let [tag (if (-> path peek int?)
                (-> path pop peek name (string/split #"s") first keyword)
                (peek path))
          path-arr (mapv strv path)
          val-str (if (map? value)
                    (-> value pr-str (string/replace #"(?<!\\), " ",\n   "))
                    (pr-str value))]
      (str tag " " val-str "\n"
           "  at path: " (string/join " > " path-arr))))

(defn value-str-id
  "Custom value string for duplicate ID error messages. Takes the form:
  > Duplicate id: <identifier>
  >  with count: <int>"
  [_ _ path value]
  (str "Duplicate id: " (-> path last strv) "\n"
       " with count: " (strv value)))

(defn value-str-ver
  "Custom value string for inScheme error messages. Takes the form:
  > Invalid inScheme: <in-scheme>
  >  at object: <identifier>
  >  profile version ids:
  >   <version-id-1>
  >   <version-id-2>
  >   ..."
  [_ _ _ value]
  (str "Invalid inScheme: " (-> value :inScheme strv) "\n"
       " at object: " (-> value :id strv) "\n"
       " profile version ids:\n  "
       (->> value :version-ids sequence sort reverse (string/join "\n  "))))

(defn value-str-edge
  "Custom value string for IRI link error messages. Takes the form:
  > Invalid <property> identifier:
  >  <identifier>
  > at object:
  >  <object map>
  > linked object:
  >  <object map>"
  [_ _ _ value]
  (let [attrs-list
        (m/match [value]
          ;; Patterns
          [{:src-primary _ :src-indegree _ :src-outdegree _ :dest-property _}]
          (str " at object:\n"
               "  {:id " (-> value :src strv) ",\n"
               "   :type " (-> value :src-type strv) ",\n"
               "   :primary " (-> value :src-primary strv) ",\n"
               "   ...}\n"
               "\n"
               " linked object:\n"
               "   {:id " (-> value :dest strv) ",\n"
               "    :type " (-> value :dest-type strv) ",\n"
               "    :" (-> value :dest-property strv) " ...,\n"
               "    ...}\n"
               "\n"
               ;; Add text that says:
               ;; "pattern is used # time(s) in the profile
               ;; and links out to # other object(s)."
               " pattern is used " (-> value :src-indegree strv) " "
               (pluralize "time" (-> value :src-indegree)) " in the profile\n"
               " and links out to " (-> value :src-outdegree strv) " other "
               (pluralize "object" (-> value :src-outdegree)) ".")
          ;; Concepts and Templates
          [{:src-version _ :dest-version _}]
          (str " at object:\n"
               "  {:id " (-> value :src strv) ",\n"
               "   :type " (-> value :src-type strv) ",\n"
               "   :inScheme " (-> value :src-version strv) ",\n"
               "   ...}\n"
               "\n"
               " linked object:\n"
               "  {:id " (-> value :dest strv) ",\n"
               "   :type " (-> value :dest-type strv) ",\n"
               "   :inScheme " (-> value :dest-version strv) ",\n"
               "   ...}")
          ;; Shouldn't happen but just in case
          :else "")]
    (str "Invalid " (-> value :type strv) " identifier:\n"
         " " (-> value :dest strv) "\n"
         "\n" attrs-list)))

(defn value-str-scc
  "Custom value string for strongly connected component error messages (if a
  digraph has a cycle). Takes the form:
  > Cycle detected in the following nodes:
  >   <identifier>
  >   <identifier>"
  [_ _ _ value]
  (str "Cycle detected involving the following nodes:\n  "
       (->> value sort (string/join "\n  "))))

;; TODO Possibly make custom error messages for @context errors?
(defn custom-printer
  "Returns a printer based on the error-type argument. A nil error-type will
  result in the default Expound printer (except with :print-specs? set to 
  false). error-types:
  - id: Duplicate ID errors
  - in-scheme: InScheme property errors
  - edge: Concept, Template and Pattern link errors
  - scc: Pattern cycle errors
  - no arg: Basic syntax + @context validation"
  [& [error-type]]
  (let [error-type (if (nil? error-type) :else error-type)]
    (case error-type
      "id"
      (exp/custom-printer {:value-str-fn value-str-id :print-specs? false})
      "in-scheme"
      (exp/custom-printer {:value-str-fn value-str-ver :print-specs? false})
      "edge"
      (exp/custom-printer {:value-str-fn value-str-edge :print-specs? false})
      "scc"
      (exp/custom-printer {:value-str-fn value-str-scc :print-specs? false})
      :else
      default-printer
      #_(exp/custom-printer {:value-str-fn value-str-def :print-specs? false}))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Property ordering in error messages 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def property-order
  {;; Profile metadata
   :id          0
   :type        1
   :inScheme    2
   :_context    3
   :conformsTo  4
   :prefLabel   5
   :definition  6
   :deprecated  7
   :author      8
   :versions    9
   :concepts    10
   :templates   11
   :patterns    12
   ;; Profile version objects
   :generatedAtTime 13
   :wasRevisionOf   14
   ;; Profile author objects
   :name  15
   :url   16
   ;; Concept properties
   :broader 17
   :broadMatch 18
   :narrower 19
   :narrowMatch 20
   :related 21
   :relatedMatch 22
   :exactMatch 23
   ;; Extension and Resource properties
   :recommendedActivityTypes  24
   :recommendedVerbs          25
   :context                   26
   :contentType               27
   :schema                    28
   :inlineSchema              29
   ;; Activity properties
   :activityDefinition        30
   :description               31
   :moreInfo                  32
   :interactionType           33
   :correctResponsesPattern   34
   :choices     35
   :scale       36
   :source      37
   :target      38
   :steps       39
   :extensions  40
   ;; Statement Template properties
   :verb                          41
   :objectActivityType            42
   :contextGroupingActivityType   43
   :contextParentActivityType     44
   :contextOtherActivityType      45
   :contextCategoryActivityType   46
   :attachmentUsageType           47
   :objectStatementRefTemplate    48
   :contextStatementRefTemplate   49
   :rules                         50
   ;; Statement Template rules properties
   :location  51
   :selector  52
   :presence  53
   :any       54
   :all       55
   :none      56
   :scopeNote 57
   ;; Pattern properties
   :primary    58
   :alternates 59
   :optional   60
   :oneOrMore  61
   :sequence   62
   :zeroOrMore 63})

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Error map manipulation functions
;;
;; XXX: Expound can act pretty funny when it comes to s/keys and s/coll-of
;; specs. See Issue #165: Improve grouping of spec errors on Expound's Github.
;; This is why we need the functions expound-error-list and expound-error-map
;; and the helper functions in this section.
;;
;; TODO: Issue #165 has been marked as a bug so it should be fixed at some
;; point. When that happens, rework this namespace to rely on Expound's native
;; error grouping functionality so it doesn't have to do awkward map manips. 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn compare-properties
  "Get the order of two properties based on how it was listed in the xAPI 
  Profile spec. Properties not listed in the property-order map will be pushed 
  to the end of the error list and sorted alphabetically.
  - neg number: prop1 comes before prop2
  - pos number: prop1 comes after prop2
  - zero: prop1 and prop2 are equal"
  [prop1 prop2]
  (let [order1 (get property-order prop1)
        order2 (get property-order prop2)]
    (cond
      (and (nil? prop1) (nil? prop2)) 0
      (and (nil? prop1) (some? prop2)) -1
      (and (some? prop1) (nil? prop2)) 1
      (and (some? order1) (nil? order2)) -1
      (and (nil? order1) (some? order2)) 1
      (and (some? order1) (some? order2)) (compare order1 order2)
      (and (nil? order1) (nil? order2)) (compare prop1 prop2))))

(defn compare-arrs
  "Compare two arrays of keywords (or strings), such that lexiographical order
  of corresponding entries takes precedence over array length.
  The conventional compares array length first, then lexicographical order.
  - Sort using compare-arrs: ([:a] [:a :b] [:z])
  - Sort using regular compare: ([:a] [:z] [:a :b])"
  [arr1 arr2]
  (loop [a1 arr1
         a2 arr2]
    (if (and (empty? a1) (empty? a2))
      0
      (let [comp-int (compare-properties (first a1) (first a2))]
        (if (= 0 comp-int)
          (recur (rest a1) (rest a2))
          comp-int)))))

(defn group-by-in
  "Splits up a spec error map by the :in path of an ::s/problems map, which
  lists the keys that lead to the value in a map data structure. For example,
  an invalid version ID would have the path ::v/version ::v/id. Returns a
  sequence of spec error maps (each with ::s/problems, ::s/spec and ::s/value
  keys) group together by :in paths.

  This is required because otherwise Expound will group together random spec
  errors together in ways that don't make sense. By doing this, we group spec
  errors by where they occur in the data structure."
  [error-map]
  (let [{problems ::s/problems spec ::s/spec value ::s/value} error-map
        problems-map (group-by :in problems)]
    (reduce-kv (fn [err-list k v]
                 (conj err-list (assoc {}
                                       ::s/problems (if (vector? v)
                                                      (sequence v) (list v))
                                       ::s/spec spec
                                       ::s/value value))) (list) problems-map)))

(defn sort-by-path
  "Sort a sequence of spec error maps by the value of the :path key, which is
  a list of all keywords that lead up to the final spec. (Unlike :key, this
  include tagged portions of specs, eg. of s/or.) Useful for keeping specs
  in the same area grouped together"
  [error-map-list]
  (sort-by #(-> % ::s/problems first :path) compare-arrs error-map-list))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Expounding functions
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn expound-error
  "Print an Expounded error message using a custom printer. The custom printer
  is determined using the error-type arg; if not supplied, it prints a default
  Expound error message (without the Relevant Specs trace)."
  [error-map & [error-type]]
  (let [print-fn (custom-printer error-type)]
    (print-fn error-map)))

(defn expound-error-list
  "Sort a list of spec error maps using the value of the :path key."
  [error-list & [error-type]]
  (doseq [error error-list]
    (do (expound-error error error-type) (println))))

(defn expound-error-map
  "Regroup an error map into a sorted list of error maps.
  Used to avoid Issue #165 for Expound."
  [error-map & [error-type]]
  (-> error-map group-by-in sort-by-path (expound-error-list error-type)))

(defn expound-errors
  "Print errors from profile validation using Expound. Available keys:
  :syntax-errors - Basic syntax validation (always present)
  :id-errors - Duplicate ID errors 
  :in-scheme-errors - inScheme property validation
  :concept-errors - Concept relation/link errors
  :template-errors - Template relation/link errors
  :pattern-errors - Pattern relation/link errors
  :pattern-cycle-errors - Cyclical pattern errors
  :context-errors - Errors in any @context maps
  :context-key-errors - Errors in expanding any keys via @context maps"
  [{:keys [syntax-errors
           id-errors
           in-scheme-errors
           concept-errors
           template-errors
           pattern-errors
           pattern-cycle-errors
           context-errors
           context-key-errors]}]
  (do
    (if (some? syntax-errors) (expound-error-map syntax-errors))
    (if (some? id-errors) (expound-error id-errors "id"))
    (if (some? in-scheme-errors) (expound-error in-scheme-errors "in-scheme"))
    (if (some? concept-errors) (expound-error-map concept-errors "edge"))
    (if (some? template-errors) (expound-error-map template-errors "edge"))
    (if (some? pattern-errors) (expound-error-map pattern-errors "edge"))
    ;; Context errors are already in list format
    (if (some? context-errors) (expound-error-list context-errors))
    (if (some? context-key-errors) (expound-error-list context-key-errors))))
